File: VB/DevExpress.Upload/DetailTemplate.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Web.UI
Imports DevExpress.Web.ASPxEditors
Imports DevExpress.Web.ASPxGridView
Imports DevExpress.Web.ASPxHtmlEditor
Imports DevExpress.Web.Data

Namespace DevExpress.Upload
	Public Class DetailTemplate
		Implements ITemplate

		Protected Friend Sub New()
		End Sub

		Public Sub InstantiateIn(ByVal container As Control) Implements ITemplate.InstantiateIn
			Dim dataRow = TryCast((TryCast(container, GridViewDetailRowTemplateContainer)).DataItem, WebCachedDataRow)
			Dim fileType = CType(dataRow("FileType"), FileType)
			Dim name = TryCast(dataRow("Name"), String)
			Dim editor = GetDetailControl(name, fileType)
			container.Controls.Add(editor)
			If fileType = FileType.Html OrElse fileType = FileType.Text Then
				Dim id = CInt(Fix(dataRow("ID")))
				Dim btnUpload = CreateUploadButton(id)
				container.Controls.Add(btnUpload)
			End If
		End Sub

		Private Function CreateUploadButton(ByVal id As Int32) As Control
			Dim button = New ASPxButton() With {.ID = "btnUpload", .Text = "Save", .AutoPostBack = False}
			button.ClientSideEvents.Click = String.Format("function() {{ grid.PerformCallback(""upload-file:{0}""); }}", id)
			Return button
		End Function

		Private Function GetBinaryImage(ByVal name As String) As Control
			Return New ASPxBinaryImage() With {.ID = "Editor", .ContentBytes = LocalHelper.ReadFileAsBytes(name)}
		End Function

		Private Function GetHtmlEditor(ByVal name As String) As Control
			Return New ASPxHtmlEditor() With {.ID = "Editor", .Html = LocalHelper.ReadFileAsString(name)}
		End Function

		Private Function GetMemo(ByVal name As String) As Control
			Return New ASPxMemo() With {.ID = "Editor", .Height = 200, .Width = 250, .Text = LocalHelper.ReadFileAsString(name)}
		End Function

		Private Function GetDetailControl(ByVal name As String, ByVal fileType As FileType) As Control
			Select Case fileType
				Case FileType.Image
					Return GetBinaryImage(name)
				Case FileType.Html
					Return GetHtmlEditor(name)
				Case FileType.Text
					Return GetMemo(name)
				Case Else
					Return Nothing
			End Select
		End Function
	End Class
End Namespace

File: CS/DevExpress.Upload/DetailTemplate.cs
using System;
using System.Web.UI;
using DevExpress.Web.ASPxEditors;
using DevExpress.Web.ASPxGridView;
using DevExpress.Web.ASPxHtmlEditor;
using DevExpress.Web.Data;

namespace DevExpress.Upload {
    public class DetailTemplate : ITemplate {

        protected internal DetailTemplate() {
        }

        public void InstantiateIn(Control container) {
            var dataRow = (container as GridViewDetailRowTemplateContainer).DataItem as WebCachedDataRow;
            var fileType = (FileType)dataRow["FileType"];
            var name = dataRow["Name"] as String;
            var editor = GetDetailControl(name, fileType);
            container.Controls.Add(editor);
            if (fileType == FileType.Html || fileType == FileType.Text)
            {
                var id = (Int32)dataRow["ID"];
                var btnUpload = CreateUploadButton(id);
                container.Controls.Add(btnUpload);
            }
        }

        Control CreateUploadButton(Int32 id) {
            var button = new ASPxButton() {
                ID = "btnUpload",
                Text = "Save",
                AutoPostBack = false
            };
            button.ClientSideEvents.Click = String.Format("function() {{ grid.PerformCallback(\"upload-file:{0}\"); }}", id);
            return button;
        }

        Control GetBinaryImage(String name) {
            return new ASPxBinaryImage() {
                ID = "Editor",
                ContentBytes = LocalHelper.ReadFileAsBytes(name)
            };
        }

        Control GetHtmlEditor(String name) {
            return new ASPxHtmlEditor() {
                ID = "Editor",
                Html = LocalHelper.ReadFileAsString(name)
            };
        }

        Control GetMemo(String name) {
            return new ASPxMemo() {
                ID = "Editor",
                Height = 200,
                Width = 250,
                Text = LocalHelper.ReadFileAsString(name)
            };
        }

        Control GetDetailControl(String name, FileType fileType) {
            switch (fileType) {
                case FileType.Image: return GetBinaryImage(name);
                case FileType.Html: return GetHtmlEditor(name);
                case FileType.Text: return GetMemo(name);
                case FileType.Unknown:
                default: return null;
            }
        }
    }
}